[BE] Team16_hiro step1_v2

hiro032 opened the pull request:
안녕하세요 16팀 백엔드를 맡은 히로 입니다. 
프론트분과 함께 현재까지 진행상황을 정리해서 pr보내봅니다.

###  pr을 보낸 브랜치가 잘못된거같아서 다른 브랜치에서 다시 pr을 보냅니다
[배포링크](http://3.34.187.250:8080/) -> index.html 정적 파일을 하나 만들어 두었습니다.

- 프론트분들과 이야기 해서 보드가 카드를 갖고있게끔 데이터를 보내주면 좋을거같다고 하셔서 맞추어서 설계를 했습니다.
- 또 하다보니 spring data jdbc를 학습하는 과정에서 aggregate와 affregate root를 학습하면서 예제를 따라서  적용하려 시도해 보았습니다

- 현재 구현한 기능은 보드 생성, 카드 생성, 카드 삭제, 보드 삭제, 카드 수정, 카드 출력, 보드 출력 입니다.
1. 보드 생성 : @Post("/boards")
2. 보드 삭제 : @Delete("/boards/{id}")
3. 보드 요청 : @Get("/boards")
4. 보드 요청 : @Get("/boards/{id}")
5. 카드 생성 : @Post("{boardId}/cards")
6. 카드 수정 : @Put("{boardId}/cards/{cardId}")
7. 카드 삭제 : @Delete("{boardId}/cards")


- api url요청 형식을 설계했습니다.
- board에 관한 요청은 (boards/)로 시작하게 됩니다.
- card에 관한 요청은 @RequestMapping({boardId}/cards)로 두어서 항상 boardId를 함께 요청하도록 설계했습니다.
- 카드를 생성하면서 저장한 후에 생성된 카드를 json으로 리턴합니다
- 초기 화면에 필요한 board와 board에 존재하는 card들을 json형식으로 모두 출력합니다.
- boardId만으로 조회시에 해당 id의 보드가 가진 카드를 출력합니다.

아직 로그인 기능과, 에러처리 부분, 카드의 로그?기록? 부분은 구현하지 못했습니다. 리뷰 해주신 부분을 수정하고 2차 pr시에 구현해보도록 하겠습니다.
감사합니다.

hiro032 committed:
init

hiro032 committed:
init

hiro032 committed:
be

hiro032 committed:
test 코드 삭제

hiro032 committed:
접근 제어자 수정

hiro032 committed:
log service제거

hiro032 committed:
웰컴페이지 생성

hiro032 committed:
welcome page 수정

hiro032 committed:
gitignore추가

hiro032 committed:
충돌 파일 수정

ghost commented:
@ksundong 충돌 수정 하는 방법 알아낸다음 수정해보겠습니다! 어디서 발생한건지 모르겠어서 시간이 조금 걸릴거같아요

ksundong commented:
이제부터는 PR Close하지 말아주세요~
아마 다른 쪽에서 먼저 머지가 되어 그런 것 같습니다. 한 번 해결해보시죠 ㅎㅎ
이쪽 저장소의 내용을 머지해주시면 됩니다.

ghost commented:
> 이제부터는 PR Close하지 말아주세요~
> 아마 다른 쪽에서 먼저 머지가 되어 그런 것 같습니다. 한 번 해결해보시죠 ㅎㅎ
> 이쪽 저장소의 내용을 머지해주시면 됩니다.

넵

ksundong commented on BE/todo-list/.gitignore:
```diff
@@ -0,0 +1,148 @@
+
+### macOS ###
+# General
+.DS_Store
```
이 파일이 빠지지 않은 이유는 무엇일까요?
git rm --cached에 대해서 한 번 공부해봅시다.

hiro032 replied:
로컬에서 삭제를 하고 푸쉬를 해도 원격에서 삭제가 되는건 아니군요!
--cached옵션으로 원격에 올라간 파일도 삭제 했습니다.

ksundong commented on BE/todo-list/build/reports/tests/test/classes/team_16.todolist.repository.CardRepositoryTest.html:
```diff
@@ -0,0 +1,354 @@
+<!DOCTYPE html>
```
이 파일도 포함되면 안되겠죠?

hiro032 replied:
--cached옵션으로 파일 제거 했습니다!

ksundong commented on BE/todo-list/build/resources/main/schema.sql:
```diff
@@ -0,0 +1,17 @@
+# DROP TABLE IF EXISTS card;
+# DROP TABLE IF EXISTS board;
```
이건 주석처리 해주신 이유가 있나요~?

hiro032 replied:
쿼리문 연습겸 계속 테이블생성하고 작성하고 하면서 넣어놨던 스키마 문장인데 배포하면서 지우지 못했네요
삭제 했습니다!

ksundong commented on BE/todo-list/build/resources/main/schema.sql:
```diff
@@ -0,0 +1,17 @@
+    title varchar(50),
+    content varchar(500),
+    created_time datetime,
+    board bigint,
+    board_key int,
```
오 spring data jdbc 떄문에 이렇게 해주신건가요?

hiro032 replied:
board_key부분은 이전에 card를 보드에 list로 관리했는데 그때 인덱스를 기록하기위해 생성한 컬럼인데 , 
최종적으로 set으로 하면서 필요가 없어진 컬럼이라 삭제 했습니다!

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/BoardController.java:
```diff
@@ -0,0 +1,48 @@
+    @PostMapping
+    public void createBoard(String title) {
+        Board board = new Board();
+        board.title = title;
```
인자에 직접 접근하게 하지 말고, 생성자를 통해서 생성해볼까요?

hiro032 replied:
title을 매개변수로 받는 생성자를 하나 추가로 생성해보았습니다.

hiro032 replied:
Board클래스의  createCard()부분도 비슷한데 Card또한 생성자를 통해서 생성하게 하는게 바람직 할까요?

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/BoardController.java:
```diff
@@ -0,0 +1,48 @@
+    @GetMapping
+    public List<Board> getBoards() {
+
+        return (List<Board>) boardRepository.findAll();
```
이렇게 형변환 해서 리턴해주는 것은 좋지 못하다고 생각됩니다.
더 좋은 방법은 없을까요?

hiro032 replied:
repository에 findAll을 오버라이드 해서 형변환 없이 값을 리턴 하게 했습니다.

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/BoardController.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @GetMapping("/{id}")
+    public ResponseEntity<?> getBoard(@PathVariable Long id) {
```
와일드카드로 무엇이든 리턴할 수 있으면 좋은걸까요? 좋지 않은걸까요?

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/BoardController.java:
```diff
@@ -0,0 +1,48 @@
+
+    @GetMapping("/{id}")
+    public ResponseEntity<?> getBoard(@PathVariable Long id) {
+        Set<Card> boards = boardRepository.findById(id).get().getCards();
```
- boards 라는 이름이 맞나요?
- `Optional`을 `get()`으로 강제 언래핑하는 것은 무조건 하지 말아주세요!
- `getCards()`로 cards를 꺼내야 할 이유는 무엇이었나요?

hiro032 replied:
- cards를 boards라고 했네요... cards로 변경했습니다
- orElseThrow로 꺼냈습니다 추후 익셉션 처리도 해보겠습니다
- 프론트 분들이 boardId값을 통해 하나의 보드가 가진 카드들을 꺼낼수있다면 좋을거같다고 하셔서 만들었습니다..

ksundong replied:
그럼 Board를 가져오는건 아니지 않을까요?
좀 더 고민해봅시다 ㅎㅎ

hiro032 replied:
getCardsByBoardId라는 이름으로 수정해 보았습니다.

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/BoardController.java:
```diff
@@ -0,0 +1,48 @@
+
+    @DeleteMapping("/{id}")
+    public void deleteBoard(@PathVariable Long id) {
+        boardRepository.deleteById(id);
```
우리 지금까지 레이어 분리를 공부했었죠?
Service Layer와 DTO를 꼭 도입해봅시다.

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,47 @@
+import team_16.todolist.repository.BoardRepository;
+
+@RestController
+@RequestMapping("/{boardId}/cards")
```
이 경로 맞는지 다시 한 번 확인해주세요!

hiro032 replied:
RequestMapping부분 말씀이신가요?? boardId를 앞에 받는게 이상한건지 어떠한 경로를 말씀하시는건지 잘 모르겠습니다 ㅠ

ksundong replied:
`boards/{boardId}/cards` 가 더 맞겠죠?

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,47 @@
+        this.boardRepository = boardRepository;
+    }
+
+    @PostMapping()
```
다른곳과 코딩 컨벤션을 맞춰주세요.

hiro032 replied:
()부분 제거했습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,47 @@
+
+    @PostMapping()
+    public Card createdCard(@PathVariable Long boardId, String title, String content) {
+        Board board = boardRepository.findById(boardId).orElseThrow(IllegalArgumentException::new);
```
옵셔널은 이렇게 써주세요 👍

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,47 @@
+        return newCard;
+    }
+
+    /*
+    update 테스트 실패?
+     */
```
어떤 주석인지 이해하기가 어렵네요.

hiro032 replied:
post맨으로 혼자 테스트를 하다가 작성한 부분인데 지우지 않고 올렸네요 제거 했습니다.

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Board.java:
```diff
@@ -0,0 +1,78 @@
+
+public class Board {
+
+    @Id
+    private Long id;
+    public String title;
```
```suggestion
    @Id
    private Long id;
    
    public String title;
```

hiro032 replied:
애노테이션이 있는 멤버는 한줄띄도록! 주의하겠습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Board.java:
```diff
@@ -0,0 +1,78 @@
+    @Id
+    private Long id;
+    public String title;
+    private Set<Card> cards = new HashSet<>();
```
Set을 사용해주신 이유가 있나요~?

hiro032 replied:
Set을 처음에 사용하고 그다음에 List로 바꿨다가 최종적으로 Set 을 사용했는데요,  List를 사용하다보니 인덱스를 별도로 관리해야하는 점이 불편해보여서 Set을 사용하게 되었습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Board.java:
```diff
@@ -0,0 +1,78 @@
+        this.title = title;
+    }
+
+    public void setCards(Set<Card> cards) {
+        this.cards = cards;
+    }
```
이 메서드가 과연 필요할지... 생각하게되네요.

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Board.java:
```diff
@@ -0,0 +1,78 @@
+        return newCard;
+    }
+
+    private Card createCard(String title, String content) {
+        Card card = new Card();
+        card.title = title;
+        card.content = content;
+        card.createdTime = LocalDateTime.now();
+        return card;
+    }
```
이건 충분히 Card 생성자가 해줄 수 있겠죠?

hiro032 replied:
위에 createBoard와 같은 부분이네요 생성자를 통해서 구현했습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Board.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public void deleteCard(Long cardId) {
+        for(Card card : cards) {
+            if(card.getId() == cardId) {
+                cards.remove(card);
+            }
```
card 자체를 받으면 for문과 if문을 우리가 써 줄 필요가 없을텐데요!

ksundong replied:
그리고 포매팅 신경써주세요!

hiro032 replied:
- 프론드에서 card객체를 받을수있다는걸 몰라서 id 값을 받아서 해결하는 걸로 구현했습니다ㅜ 다시 질문드려보니 가능할거같다고 하셔서 card객체를 받는 식으로 변경했습니다
- 포메팅도 수정하겠습니다!

ksundong replied:
id로 받아도, card를 구해오면 되겠죠?

hiro032 replied:
혹시 그럼 cardId를 통해 그에 해당하는 card객체를 받는 메서드를 생각해보라는 말씀이신가요??

ksundong replied:
네 그런 메서드가 있으면 Board는 Card를 사용해서 무엇인가 할 수 있겠죠?

hiro032 replied:
확실히 cardId를 통해 card를 받는 메서드를 만들었더니 많은 곳에서 중복이 사라졌습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Board.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public Card updateCard(Long cardId, Card updateCard) {
+        for(Card beforeCard : cards) {
+            if(beforeCard.getId() == cardId) {
+                beforeCard.updateCard(updateCard);
+                return beforeCard;
+            }
+        }
```
```suggestion
        for (Card beforeCard : cards) {
            if (beforeCard.getId() == cardId) {
                beforeCard.updateCard(updateCard);
                return beforeCard;
            }
        }
```

hiro032 replied:
넵 주의하겠습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Board.java:
```diff
@@ -0,0 +1,78 @@
+                return beforeCard;
+            }
+        }
+        return null;
```
해당 card가 없으면 못찾았다고 Exceptino이 나야 더 정상이라고 생각됩니다.
그리고 `null`을 리턴하는 것 역시 지양해주세요.

ksundong replied:
한 번 코드를 봐야겠네요~

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Board.java:
```diff
@@ -0,0 +1,78 @@
+        return "Column{" +
+                "id=" + id +
+                ", title='" + title + '\'' +
+                ", cards=" + cards +
```
그냥 cards를 toString에 작성해버리면, 오류가 발생할 포인트가 될 수 있습니다.

hiro032 replied:
toString을 사용하지 않아서 제거 했습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Log.java:
```diff
@@ -0,0 +1,15 @@
+
+public class Log {
+
+    @Id
+    private Long id;
+    private String  title;
```
```suggestion
    @Id
    private Long id;
    
    private String title;
```

hiro032 replied:
수정 완료 했습니다!

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Log.java:
```diff
@@ -0,0 +1,15 @@
+    private String  title;
+    private LocalDateTime time;
+    private String prevBoard;
+    private String currBoard;
```
가급적이면 단축어보다는 풀네임을 적어주세요.

hiro032 replied:
아직 프론트분들과 상의 하지 않아서 임의로 썼었는데요, 
- previousBoard
- currentBoard 로 결정했습니다

ksundong replied:
넵넵

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/repository/BoardRepository.java:
```diff
@@ -0,0 +1,11 @@
+import team_16.todolist.domain.Board;
+
+
+@Repository
```
굳이 안 붙여줘도 되는 애노테이션입니다.

hiro032 replied:
넵

ksundong commented on BE/todo-list/src/main/resources/application.properties:
```diff
@@ -0,0 +1,9 @@
+spring.datasource.password=2230
+spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
+spring.datasource.schema=classpath:schema.sql
+spring.datasource.initialization-mode=always
```
동일한 파일이 여러개가 있는 것 같은데 확인 한 번 해주세요.

ksundong commented on BE/todo-list/src/main/resources/static/index.html:
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>Title</title>
+</head>
+<body>
+team16<br>
+todo-list
+</body>
+</html>
```
삭제해주셔도 괜찮습니다.

hiro032 replied:
넵

ksundong reviewed (changes requested):
안녕하세요! Hiro!! 리뷰어 Dion입니다!!
첫 프로젝트를 수행하느라 대단히 고생많으셨습니다. 팀 프로젝트가 처음이니 만큼 여러가지 의사소통 할 일도 많으셨을거에요.
더군다나 git conflict 해결이 쉽지 않은 문제였어서 많은 고민을 하셨을 것 같습니다.

코드는 조금 아쉬운 부분이 보여서 변경요청을 드렸습니다. 차근차근 반영해주세요.

좀 더 힘내셔서 이번 미션에 많이 배워가셨으면 좋겠습니다. 😊

hiro032 committed:
index.html파일 삭제

hiro032 committed:
불필요한 주석 제거

hiro032 committed:
불필요한 파일 삭제

hiro032 committed:
schema.sql 파일 주석제거

hiro032 committed:
BoardController의 getBoard메서드 부분 수정

boards를 board라는 변수명으로 수정, Optional을 get()으로 꺼낸 부분을 orElseThrow()로 수정

hiro032 committed:
Board의 생성자 추가

디폴트 생성자에 title만을 매개변수로 받는 생성자 추가

hiro032 committed:
Card의 생성자 추가

Board클래스의 createCard부분에서 card를 생성할때 사용하는 title, content ,createdTime을 매개변수로 가진 Card생성자 추가

hiro032 committed:
Board 클래스의 deleteCard부분 수정

cardId를 받아와서 for문을 통해 cards에서 삭제를 했지만, card객체를 받아서 for문 없이 card를 삭제하는 방법으로 수정

hiro032 committed:
Board 클래스의 toString제거
11

hiro032 committed:
Board클래스의 updateCard부분에서 Optional처리

hiro032 committed:
BoardService구현

hiro032 committed:
LogRepository생성

hiro032 committed:
수정사항 수정

hiro032 committed:
getBoard -> getCardsByBoardId로 수정

hiro032 committed:
오타 수정

ksundong reviewed (commented):
gitignore에 ignore된 전체 파일을 `git rm --cached` 해주신 다음에 다시 `git add`를 하셔서 올려주세요.
협업을 할 때에는 자동생성되는 파일은 버전관리 하지 않습니다.

ksundong commented on BE/todo-list/build/resources/main/schema.sql:
```diff
@@ -0,0 +1,26 @@
+    time datetime,
+    previous_board varchar(50),
+    current_board varchar(50)
+)
```
통행금지 표시가 나오는데 관련 설정이 되어있나요?

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/BoardController.java:
```diff
@@ -0,0 +1,42 @@
+        return boardService.getBoards();
+    }
+
+    @GetMapping("/{id}")
```
`/{id}/cards`가 되는 것이 맞지 않을까요?

hiro032 replied:
수정 했습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,54 @@
+
+    /*
+    Card객체를 매개변수로 받는것으로 수정 예정입니다.
+    포스트맨 으로 테스트하기위해 title과 content로 받습니
```
포스트맨하고는 전혀 관계 없습니다 ㅎㅎ card로 받아보세요.

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,54 @@
+    @PostMapping
+    public Card createdCard(@PathVariable Long boardId, String title, String content) {
+        Card createdCard = boardService.createCard(boardId, title, content);
+        logService.createLog(Log.Action.ADD, createdCard.getId(), boardId);
```
controller에서 log를 남기는 것이 아니라, 해당 service에서 logService를 호출하는 식의 설계가 더 적절하다고 봅니다.

ksundong replied:
트랜잭션하고도 관련이 어느정도 있는데요, 한 번 생각해보세요.

hiro032 replied:
- 컨트롤러에서 로그 생성을 호출하지 않고, 보드 서비스에서 로그 서비스의 생성을 호출하는 방식으로 변경해보았습니다
- 트랜잭션도 추가 학습을 해보겠습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,54 @@
+
+    @PutMapping("/{cardId}")
+    public Card updateCard(@PathVariable Long boardId, @PathVariable Long cardId, String title, String content) {
+        logService.createLog(Log.Action.UPDATE, cardId, boardId);
+
+        return boardService.updateCard(boardId, cardId, title, content);
```
동작의 순서도 신경써주세요~

hiro032 replied:
서비스로 이동하면서, 생성을 먼저하고, 그 이후에 로그를 생성하는 식으로 순서를 바꾸어 봤습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/CardController.java:
```diff
@@ -0,0 +1,54 @@
+        logService.createLog(Log.Action.REMOVE, cardId, boardId);
+        boardService.deleteCard(boardId, cardId);
+    }
+
+
```
빈 라인도 신경써주세요~

hiro032 replied:
넵 수정하고 마지막 개행은 한줄만 남겨두는걸로 했습니다.

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/LogController.java:
```diff
@@ -0,0 +1,25 @@
+@RestController
+@RequestMapping("/logs")
+public class LogController {
+    private final LogService logService;
```
다른 코드와 필드 선언된 위치가 다른데요. 이것도 역시 컨벤션이라고 볼 수 있습니다. 확인하시고, 변경해주세요!

hiro032 replied:
넵 감사합니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/controller/LogController.java:
```diff
@@ -0,0 +1,25 @@
+
+    @GetMapping
+    public Set<Log> getLog() {
+
```
의미없는 빈 라인은 제거를 해주시는게 좋을 것 같습니다.
다른 부분도 살펴봐주세요.

hiro032 replied:
return문위에는 한칸 띄는건줄알았습니다 ㅜ
빈 라인은 제거 하겠습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Board.java:
```diff
@@ -0,0 +1,76 @@
+        return newCard;
+    }
+
+    private Card createCard(String title, String content) {
+        Card card = new Card(title, content);
+
+        return card;
+    }
+
```
그냥 생성자를 쓰는것과 어떤 차이가 있나요?

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Board.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    public void deleteCard(Long cardId) {
+        getCard(cardId);
```
board 안에 getCard가 있는것은 제가 의도했던 바는 아니지만, 잘 동작한다면 되었습니다. ㅎㅎ
근데 문제가 될 수도 있어보여요~

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Board.java:
```diff
@@ -0,0 +1,76 @@
+
+    public void deleteCard(Long cardId) {
+        getCard(cardId);
+        cards.remove(getCard(cardId));
```
위에 getCard를 변수로 뽑아낸다면, 굳이 getCard를 두 번 호출하지 않아도 되겠죠?

hiro032 replied:
전혀 필요없는 코드 였네요 ㅜ 수정했습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Board.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    public Card getCard(Long cardId) {
+        Optional<Card> findCard = null;
```
`Optional`을 `null`로 초기화하지 말아주세요.
`null`인 `Optional` 객체를 만들려면 그냥 `Optional.empty()`를 써주세요.

hiro032 replied:
empty()가 있는줄은 몰랐네요, 감사합니다! 수정했습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Board.java:
```diff
@@ -0,0 +1,76 @@
+        return getCard(cardId).updateCard(title, content);
+    }
+
+    public Card getCard(Long cardId) {
+        Optional<Card> findCard = null;
+
+        for (Card card : cards) {
+            if (card.getId() == cardId) {
+                findCard = Optional.of(card);
+            }
+        }
+
+        return findCard.orElseThrow(IllegalArgumentException::new);
+    }
```
stream을 사용하면 간단하게 표현할 수 있습니다.
```suggestion
       return cards.stream().filter(card → card.getId() == cardId).findOne().orElseThrow(IllegalArgumentException::new);
```

그리고, IllegalArgumentExcpetion 대신에 CustomException의 사용을 고려해봐주세요.

ksundong replied:
줄바꿈은 안했는데, 줄바꿈을 해주면 더 가독성이 높아지겠죠?

hiro032 replied:
- 스트림 사용이 익숙치 않았는데 한번더 배웠습니다ㅜ
- 스트림 적용후 줄바꿈 추가했습니다
- 커스텀익셉션을 다음에 한번에 추가하도록 해보겠습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Card.java:
```diff
@@ -0,0 +1,64 @@
+    @Id
+    private Long id;
+
+    String title;
+    String content;
+    LocalDateTime createdTime;
```
뭔가 이상하지 않나요? 위의 id와 뭐가 다른가요?

hiro032 replied:
이전에 직접 맴버 변수에 접근에서 값을 넣어서 접근 제어자를 붙이지 않았는데 생성자로 넣어주게 수정하면서 같이 수정하지 못했네요
접근제어자 추가해주었습니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Card.java:
```diff
@@ -0,0 +1,64 @@
+package team_16.todolist.domain;
+
+import org.springframework.data.annotation.Id;
+import java.time.LocalDateTime;
+public class Card {
```
```suggestion
import java.time.LocalDateTime;

public class Card {
```

hiro032 replied:
항상 놓치네요 알려주셔서 감사합니다 ㅜ
습관이 들기가 힘드네요

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Log.java:
```diff
@@ -0,0 +1,68 @@
+    private Action action;//이동 수정 삭제 생성
+    private String cardTitle;//카드 제목
+    private LocalDateTime time;
+    private String previousBoard;//보드 이름
+    private String currentBoard;//보드 이름
```
board 이름을 넣는게 좋을까요? 아니면 board를 가지고 있는게 좋을까요?
log에는 어떤 정보가 담기는게 좋을지도 고민해봅시다.

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Log.java:
```diff
@@ -0,0 +1,68 @@
+        return currentBoard;
+    }
+
+    public enum Action {
+        ADD, MOVE, REMOVE, UPDATE;
+    }
```
별도의 클래스로 추출해주세요.

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Log.java:
```diff
@@ -0,0 +1,68 @@
+        ADD, MOVE, REMOVE, UPDATE;
+    }
+
+    private Log() {
+    }
```
이렇게 하면 JPA에서 못 쓸텐데 괜찮았나요?

hiro032 replied:
지금은 SpringData JDBC를 쓰고있어서 괜찮은거 같습니다.
- 기본생성자가 존재하는 경우 1순위로 사용되고
- 기본 생성자를 private으로 선언해도 정상작동함
log는 팩터리 메서드로 객체를 생성해서 괜찮은거같습니다..

ksundong replied:
앗 제가 착각했네요 ㅎㅎ

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Log.java:
```diff
@@ -0,0 +1,68 @@
+        this.currentBoard = currentBoard;
+    }
+
+    public static Log newLog(Action action, String cardTitle, String currentBoard) {
+        return new Log(action, cardTitle, currentBoard);
+    }
```
정적 팩토리 메서드의 사용 좋습니다.

hiro032 replied:
감사함니다

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/domain/Log.java:
```diff
@@ -0,0 +1,68 @@
+        return new Log(action, cardTitle, currentBoard);
+    }
+
+    public static Log newLogByPreviousBoard(Action action, String cardTitle, String previousBoard, String currentBoard) {
```
메서드가 말하고자 하는바가 무엇인가요? 보드이동인가요?
'이전 보드에 의해 새로운 로그 생성' 정도로 번역될 것 같은데, 좀 더 깔끔하게 바꿔보는건 어떨까요?

hiro032 replied:
네이밍 컨벤션이 있길래 따라해봤는데 의미가 이상하네요 확실히..
createLogByMoveBoard로 수정해봤습니다

ksundong replied:
By 쓰지말고, createMoveBoardLog가 훨씬 깔끔할 것 같은데요~

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/repository/BoardRepository.java:
```diff
@@ -0,0 +1,11 @@
+package team_16.todolist.repository;
+
+import org.springframework.data.repository.CrudRepository;
+import team_16.todolist.domain.Board;
```
패키지명에는 -, _는 안쓰는 것이 대부분이더라고요~ 참고해주세요!

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/service/BoardService.java:
```diff
@@ -0,0 +1,70 @@
+        boardRepository.deleteById(id);
+    }
+
+    public Card createCard(Long boardId, String title, String content) {
```
혹시 `@Transactional` 애노테이션에 대해서 학습하셨나요?
학습해보시길 권장합니다~

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/service/BoardService.java:
```diff
@@ -0,0 +1,70 @@
+        Board moveBoard = boardRepository.findById(moveBoardId).orElseThrow(IllegalArgumentException::new);
+
+        Card card = board.getCard(cardId);
+        deleteCard(boardId, cardId);
```
move를 수행하면 delete가 그냥 되게끔 개발해보는 방법도 있을 것 같네요~

ksundong replied:
너무 어려우면 나중에 해봅시다.

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/service/LogService.java:
```diff
@@ -0,0 +1,47 @@
+
+@Service
+public class LogService {
+    private final LogRepository logRepository;
+    private final BoardRepository boardRepository;
```
가급적이면 1 Service 1 Repository로 개발해보세요.

ksundong replied:
다른 Repository가 필요하면 Service를 거치는 방식도 생각해보세요.

ksundong commented on BE/todo-list/src/main/java/team_16/todolist/service/LogService.java:
```diff
@@ -0,0 +1,47 @@
+        saveLog(log);
+    }
+
+    public void createLogByPreviousBoardId(Log.Action action, Long cardId, Long previousBoardId, Long currentBoardId) {
+        Board currentBoard = boardRepository.findById(currentBoardId).orElseThrow(IllegalArgumentException::new);
+        Board previousBoard = boardRepository.findById(previousBoardId).orElseThrow(IllegalArgumentException::new);
+
+        String cardTitle = previousBoard.getCard(cardId).getTitle();
+
+        Log log = Log.newLogByPreviousBoard(action, cardTitle, currentBoard.getTitle(), previousBoard.getTitle());
+
+        saveLog(log);
+
+    }
```
복잡하죠? 이렇게 복잡하지 않아도 될 것 같은데, 어떻게 개선할 수 있을까요?

ksundong commented on BE/todo-list/build/resources/main/application.properties:
```diff
@@ -0,0 +1,9 @@
+#
+spring.datasource.url=jdbc:mysql://localhost:3306/todolist
+# MySQL 설정
+spring.datasource.username=root
+spring.datasource.password=2230
+spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
+#spring.datasource.schema=classpath:schema.sql
+spring.datasource.initialization-mode=always
+
```
이 파일은 build에 남아있는데요, 이것 역시 제거대상입니다~

ksundong commented on BE/todo-list/src/main/resources/application.properties:
```diff
@@ -0,0 +1,9 @@
+spring.datasource.username=root
+spring.datasource.password=2230
+spring.datasource.driver-class-name=com.mysql.cj.jdbc.Driver
+#spring.datasource.schema=classpath:schema.sql
```
필요없어졌다면 지워주세요~

ksundong reviewed (changes requested):
고생하셨습니다.
아직 git에서 제대로 제거되지 않은 것들이 있어서 변경요청 드리고
생각해보실 부분들에 대해서 코멘트 남겼어요. 확인부탁드려요~

ksundong commented:
gradle 관련 내용 제거해주시면 merge 할게요~~

hiro032 committed:
Controller에 있던 logService를 BoardService로 이동

hiro032 committed:
Log 생성메서드 네임 수정

hiro032 committed:
DS_Store삭제

hiro032 committed:
test

hiro032 committed:
BE폴더 삭제

hiro032 committed:
.gradle폴더 제거

hiro032 committed:
DS_Store제거

hiro032 committed:
DS_Store제거

hiro032 committed:
gitignore 수정

hiro032 committed:
gitignore

hiro032 committed:
de

hiro032 committed:
gitignore 수정
